src/librustc_typeck/check/_match.rs

```diff
@@ -626,9 +626,9 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
         let discrim_diverges = self.diverges.get();
         self.diverges.set(Diverges::Maybe);
 
-        // Typecheck the patterns first, so that we get types for all the
-        // bindings.
-        let all_arm_pats_diverge = arms.iter().map(|arm| {
+        // rust-lang/rust#55810: Typecheck patterns first (via eager
+        // collection into `Vec`), so we get types for all bindings.
```

Comment:
Ahh, had I seen this PR, I would've suggested adding a note that the `map` below has side-effects in terms of type inference (unifying the types of all the arms together), and that's why it has to fully complete first.